pyfr/integrators/multip.py

```diff
@@ -0,0 +1,260 @@
+
+    def prolongate(self, l1, l2):
+        l1idxcurr, l2idxcurr = self._mgidxcurr[l1], self._mgidxcurr[l2]
+        l1sys, l2sys = self._mgsystem[l1], self._mgsystem[l2]
+        mg0, mg1, mg2 = self._mg_regidx
+
+        # Correction with respect to the non-smoothed value from down-cycle
+        # mg1 = Delta = Q^s - Q^ns
+        self._add(0, mg1, 1, l1idxcurr, -1, mg2)
+
+        # Prolongate the correction and store to mg1
+        l1sys.eles_scal_upts_inb.active = mg1
+        l2sys.eles_scal_upts_inb.active = mg1
+        self._queue % self.mgproject(l1, l2)()
+
+        # Add the correction to the end quantity at l2
+        # Q^m+1  = Q^s + Delta
+        self.level = l2
+        self._add(1, l2idxcurr, 1, mg1)
+
+    def _rhs_with_dts(self, t, uin, fout, c=1, passmg=False):
+        # Compute -∇·f
+        self.system.rhs(t, uin, fout)
+
+        # Coefficients for the dual-time source term
+        svals = [c*sc for sc in self._dual_time_source]
+
+        # Source addition -∇·f - dQ/dt
+        axnpby = self._get_axnpby_kerns(len(svals) + 1, level=self.level,
+                                        subdims=self._subdims)
+        self._prepare_reg_banks(fout, self._idxcurr, *self._source_regidx)
+        self._queue % axnpby(1, *svals)
+
+        # Multigrid r addition
+        if self.level != 0 and not passmg:
+            axnpby = self._get_axnpby_kerns(2, level=self.level)
+            self._prepare_reg_banks(fout, self._mg_regidx[0])
+            self._queue % axnpby(1, -1)
+
+    def _add(self, *args):
+        # Get a suitable set of axnpby kernels
+        axnpby = self._get_axnpby_kerns(len(args) // 2, level=self.level)
+
+        # Bank indices are in odd-numbered arguments
+        self._prepare_reg_banks(*args[1::2])
+
+        # Bind and run the axnpby kernels
+        self._queue % axnpby(*args[::2])
+
+    @property
+    def _mg_regidx(self):
+        return self._regidx[-3:]
+
+    @property
+    def system(self):
+        return self._mgsystem[self.level]
+
+    def _init_reg_banks(self):
+        # Three additional banks for multigrid
+        self._mgregs, self._regidx = [], list(range(self.nreg + 3))
+        self._mgidxcurr = [0]*self.nlvls
+
+        # Create a proxylist of matrix-banks for each storage register
+        for l in range(self.nlvls):
+            self._mgregs.append(
+                [proxylist([self.backend.matrix_bank(em, i)
+                            for em in self._mgsystem[l].ele_banks])
+                 for i in self._regidx]
+            )
+
+    def _init_system(self, systemcls, *args):
+        self._mgsystem = [systemcls(*args, nreg=self.nreg+3, cfg=self.mgcfg[i])
+                          for i in range(self.nlvls)]
+        self._init_proj_mats()
+
+    def advance_to(self, t):
+        if t < self.tcurr:
+            raise ValueError('Advance time is in the past')
+
+        lvls = list(range(self.nlvls - 1)) + list(range(self.nlvls - 1, -1, -1))
```

Comment:
Might want to make this a class member as it does not change between iterations.